### ISSUE TITLE
Unable to load models from "Private" s3 location

### ISSUE THREAD

**Issue by alnhk** [triton-inference-server/server#5790]
This is similar to this earlier and closed issue - https://github.com/triton-inference-server/server/issues/5353 

Problem description:
 - We are using `tritonserver:23.02-py3` 
 - We are trying with "private" S3 which is on top of Pure Blade Flash Storage we trying with
 - We did exported the value of AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY only as pure blade flash storage dont provide anything else.
 - When starting, we are encountering this issue - `E0515 09:15:18.143538 2299 tritonserver.cc:2273] Internal: Unable to create S3 filesystem client. Check account credentials.` 
 - Below is the piece of the log 
 - We did stored the certificates under  /usr/local/share/ca-certificates and ran update-ca-certificates 
```

+ /opt/tritonserver/bin/tritonserver --exit-on-error=false --model-repository=s3://acme.example.com/acmebucket/model_repository --grpc-port=7001 --http-port=8071 --metrics-port=7002
I0515 09:14:16.967976 2299 pinned_memory_manager.cc:240] Pinned memory pool is created at '0x7f680e000000' with size 268435456
I0515 09:14:16.969992 2299 cuda_memory_manager.cc:105] CUDA memory pool is created on device 0 with size 67108864
I0515 09:15:18.142688 2299 metrics.cc:864] Collecting metrics for GPU 0: NVIDIA H100 1TB PCIe
I0515 09:15:18.143104 2299 metrics.cc:757] Collecting CPU metrics
I0515 09:15:18.143513 2299 tritonserver.cc:2264]
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+
| Option                           | Value                                                                                                                                              |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+
| server_id                        | triton                                                                                                                                             |
| server_version                   | 2.31.0                                                                                                                                             |
| server_extensions                | classification sequence model_repository model_repository(unload_dependents) schedule_policy model_configuration system_shared_memory cuda_shared_ |
|                                  | memory binary_tensor_data statistics trace logging                                                                                                 |
| model_repository_path[0]         | s3://acme.example.com/acmebucket/model_repository                                                               |
| model_control_mode               | MODE_NONE                                                                                                                                          |
| strict_model_config              | 0                                                                                                                                                  |
| rate_limit                       | OFF                                                                                                                                                |
| pinned_memory_pool_byte_size     | 268435456                                                                                                                                          |
| cuda_memory_pool_byte_size{0}    | 67108864                                                                                                                                           |
| response_cache_byte_size         | 0                                                                                                                                                  |
| min_supported_compute_capability | 6.0                                                                                                                                                |
| strict_readiness                 | 1                                                                                                                                                  |
| exit_timeout                     | 30                                                                                                                                                 |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+

E0515 09:15:18.143538 2299 tritonserver.cc:2273] Internal: Unable to create S3 filesystem client. Check account credentials.
I0515 09:15:18.145360 2299 grpc_server.cc:4868] Started GRPCInferenceService at 0.0.0.0:7001
I0515 09:15:18.145563 2299 http_server.cc:3477] Started HTTPService at 0.0.0.0:8071
I0515 09:15:18.186864 2299 http_server.cc:184] Started Metrics Service at 0.0.0.0:7002
```

**Comment by alnhk** [comment]
Hello @yeahdongcn : 

Thank you for the reply, Hmm, even with  Despite AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, `AWS_DEFAULT_REGION` provided along with "host:port" included, we still encounter  same issue (i.e `Internal: Unable to create S3 filesystem client. Check account credentials.`) , yes, we are using this on top of kubernetes. 

 Question mark - we have non-prod and prod based pure flash storage with S3, we are using this non-prod ones for this POC, and this one donot have valid SSL Certificate, is there a way to bypass certificate ?

**Comment by yeahdongcn** [comment]
Just pushed an example to list objects from a Minio server using `aws-sdk-go`. It could be used to verify your private s3 configuration.

Please see: https://github.com/yeahdongcn/minio

**Comment by alnhk** [comment]
> Just pushed an example to list objects from a Minio server using `aws-sdk-go`. It could be used to verify your private s3 configuration.
> 
> Please see: https://github.com/yeahdongcn/minio

Tried this, using minio compiled and generated out of `aws-sdk-go` did works well, using the same credentials, AWS_DEFAULT_REGION, endpoint with port (http/80), and bucket with prefix, it works well and it is able to print the model data according to prefix set. However, with this tritonserver, we are still hitting with `Internal: Unable to create S3 filesystem client. Check account credentials.` 
```
root@87f051207d1a:~/go/minio-main# ./minio |grep pbtxt|wc -l
4
root@87f051207d1a:~/go/minio-main#
``` 

Even with this direct docker  image `tritonserver:23.02-py3 `, it did not work using direct private S3.

**Comment by yeahdongcn** [comment]
The error is reported here:

```c
Status
S3FileSystem::CheckClient(const std::string& s3_path)
{
  std::string bucket, object_path;
  RETURN_IF_ERROR(ParsePath(s3_path, &bucket, &object_path));
  // check if can connect to the bucket
  s3::Model::HeadBucketRequest head_request;
  head_request.WithBucket(bucket.c_str());
  if (!client_->HeadBucket(head_request).IsSuccess()) {
    return Status(
        Status::Code::INTERNAL,
        "Unable to create S3 filesystem client. Check account credentials.");
  }
  return Status::Success;
}
```

Maybe you could do the following checks:

* Start triton server container by `docker run -it ... bash` and check private S3 connectivity inside the container
* Use `printenv` to check whether the `AWS_` envs are correctly passed to the container

**Comment by alnhk** [comment]
@yeahdongcn  : Yes , we confirm that AWS_ envs were correctly passed. Can you please  atleast  provide another docker image with proper loggings  to print further error using exception(e) message equivalent ?

**Comment by GuanLuo** [comment]
@kthui ^^^

**Comment by kthui** [comment]
Hi @alnhk, would you be able to try fetching the head object from an [aws s3api](https://docs.aws.amazon.com/cli/latest/reference/s3api/head-object.html) command? We have seen misconfigured permission in the past which prevented the head object from being fetched, the command can help us figure out if the issue is related to permission or something else.

In your case, the command should look something like:
```
aws s3api head-object --bucket acmebucket --key model_repository/model/config.pbtxt
```
You might also need to set some [global options](https://docs.aws.amazon.com/cli/latest/reference/s3api/head-object.html#global-options) to direct the request into your service.

**Comment by alnhk** [comment]
@kthui : 
I have no issue with running `aws s3api` 
```
root@e73e96ae6dde:~# aws s3api --endpoint-url=http://acme.example.com head-object --bucket acmebucket --key triton/model/acme/config.pbtxt
{
    "LastModified": "Wed, 17 May 2023 04:19:05 GMT",
    "ContentLength": 318,
    "ETag": "\"7406191b377832f9f4da2dc84f8c331b\"",
    "VersionId": "AAAAAAAAAAMgL9nXIPf7uQAwAAADZ13vAAAAAAAAAAEAMAAAAAMU7gCQAAAAAAABAAAAAAAAABMAZnRyaXRvbi9tb2RlbC9wZGYyb2NyL2NvbmZpZy5wYnR4dA..",
```

**Comment by alnhk** [comment]
Is there any public channel available like slack or equivalent available ?  Or is there no importance to use private S3 bucket outside AWS ?

**Comment by alnhk** [comment]
@kthui  : 
Per this  below, can you please provide the  docker image + tag detail  with this below changes ? Cause, we dont know how to rebuild (like what version of gcc, cmake etc).
https://github.com/triton-inference-server/core/pull/205/files

**Comment by seoungbae-park** [comment]
only give aws credentials as env? 
My problem is that i didn't give region env

**Comment by alnhk** [comment]
Yes,  we only give AWS creds as env, is it possible if you can re-generate docker image with "temporary" tag which contains this fix ? https://github.com/triton-inference-server/core/pull/205/files

**Comment by alnhk** [comment]
> only give aws credentials as env? My problem is that i didn't give region env

Please include region env

**Comment by kthui** [comment]
Hi alnhk, we have added a [patch](https://github.com/triton-inference-server/core/commit/58e52b3449678ac436d549374cb44d66071b7e68) that will stop the S3 client from trying to upgrade to HTTP/2. It is scheduled to be included in the upcoming 23.06 release. I will be closing the issue, but please let us know if the problem persisted with the patch.

**Comment by ducfilan** [comment]
For a local or private instance of S3, the prefix s3:// must be followed by the host and port (separated by a semicolon) and subsequently the bucket path.

$ tritonserver --model-repository=s3://host:port/bucket/path/to/model/repository ...

Fort port here it's **important**, even it's 80, need to specify it explicitly.
